core/src/test/java/org/apache/calcite/test/JdbcTest.java

```diff
@@ -3119,6 +3125,96 @@ void testInnerJoinValues(String format) {
+   * Disabling trimming of unused fields via config and program. */
+  @Test void testJoinWithTrimmingEnabledByConfig() {
+    CalciteAssert.hr().query("select \"d\".\"name\" from \"hr\".\"depts\" as \"d\" \n"
+                + "  join \"hr\".\"emps\" as \"e\" on \"d\".\"deptno\" = \"e\".\"deptno\" \n")
+        .withHook(Hook.SQL2REL_CONVERTER_CONFIG_BUILDER,
+            (Consumer<Holder<Config>>) configHolder ->
+            configHolder.set(configHolder.get().withTrimUnusedFields(true)))
+        .withHook(Hook.PROGRAM,
+            (Consumer<Holder<Program>>)
+            programHolder -> programHolder
+                    .set(sequence(Programs.SUB_QUERY_PROGRAM, getProgram(), Programs.CALC_PROGRAM)))
+        .convertContains(""
+                + "LogicalProject(name=[$1])\n"
+                + "  LogicalJoin(condition=[=($0, $2)], joinType=[inner])\n"
+                + "    LogicalProject(deptno=[$0], name=[$1])\n"
+                + "      LogicalTableScan(table=[[hr, depts]])\n"
+                + "    LogicalProject(deptno=[$1])\n"
+                + "      LogicalTableScan(table=[[hr, emps]])\n"
+                + "");
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-5984">[CALCITE-5984]</a>
+   * Disabling trimming of unused fields via config and program. */
+  @Test void testJoinWithTrimmingEnabledByProgram() {
+    CalciteAssert.hr().query("select \"d\".\"name\" from \"hr\".\"depts\" as \"d\" \n"
+                + "  join \"hr\".\"emps\" as \"e\" on \"d\".\"deptno\" = \"e\".\"deptno\" \n")
+        .withHook(Hook.SQL2REL_CONVERTER_CONFIG_BUILDER,
+            (Consumer<Holder<Config>>) configHolder ->
+            configHolder.set(configHolder.get().withTrimUnusedFields(false)))
+        .withHook(Hook.PROGRAM,
+            (Consumer<Holder<Program>>)
+            programHolder -> programHolder
+                    .set(Programs.standard()))
+        .convertContains(""
+                + "LogicalProject(name=[$1])\n"
+                + "  LogicalJoin(condition=[=($0, $2)], joinType=[inner])\n"
+                + "    LogicalProject(deptno=[$0], name=[$1])\n"
+                + "      LogicalTableScan(table=[[hr, depts]])\n"
+                + "    LogicalProject(deptno=[$1])\n"
+                + "      LogicalTableScan(table=[[hr, emps]])\n"
+                + "");
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-5984">[CALCITE-5984]</a>
+   * Disabling trimming of unused fields via config and program. */
+  @Test void testJoinWithTrimmingEnabledByProgramAndConfig() {
+    CalciteAssert.hr().query("select \"d\".\"name\" from \"hr\".\"depts\" as \"d\" \n"
+                + "  join \"hr\".\"emps\" as \"e\" on \"d\".\"deptno\" = \"e\".\"deptno\" \n")
+        .withHook(Hook.SQL2REL_CONVERTER_CONFIG_BUILDER,
+            (Consumer<Holder<Config>>) configHolder ->
+                configHolder.set(configHolder.get().withTrimUnusedFields(true)))
```

Comment:
I see the sql is same as above,the withTrimUnusedFields method has different parameters, but the plan is the same.

Why did you use this SQL example？

Reply:
Allowing field trimming happens by

1. Specifying SQL2RelConverter Config  
**or**
2. Specifying a Program that contains the Field trimming program (the standard program)

I tried to showcase all the different ways config 1 & 2 will enable trimming, while trimming will be disabled if both 1. Converter config and 2. Program Config disable it

Reply:
Hi @WegdanGhazi ,thanks for your reply.

I have saw the difference between `testJoinWithTrimmingEnabledByProgramAndConfig` and `testJoinWithTrimmingEnabledByProgram`  is the withTrimUnusedFields param value,one is false,another is true.

But the input sql and expected plan is same,so I'm a little confused about the purpose of these two unit tests.I thought a few code comments illustrating this case would make it easier for others to understand.

Reply:
Got it! Will add comments, thanks!